core/platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/MachineManager.java

```diff
@@ -803,13 +803,16 @@ private SnapshotImpl doSaveMachine(SnapshotImpl snapshot, Instance machine) thro
 
             snapshotWithKey = new SnapshotImpl(snapshot);
             snapshotWithKey.setInstanceKey(machine.saveToSnapshot(machine.getOwner()));
-            //TODO remove old snapshot
+
             try {
-                snapshotDao.removeSnapshot(snapshotDao.getSnapshot(snapshot.getWorkspaceId(), snapshot.getEnvName(), snapshot.getMachineName()).getId());
+                SnapshotImpl outOfDateSnapshot = snapshotDao.getSnapshot(snapshot.getWorkspaceId(),
+                                                                         snapshot.getEnvName(),
+                                                                         snapshot.getMachineName());
+                snapshotDao.removeSnapshot(outOfDateSnapshot.getId());
+                machineInstanceProviders.getProvider(outOfDateSnapshot.getType()).removeInstanceSnapshot(outOfDateSnapshot.getInstanceKey());
             } catch (NotFoundException ignored) {
```

Comment:
Now NotFoundException of provider can be caught here too. But comment doesn't say about that.